doc/developer/design/20230801_sink_conn_mgmt.md

```diff
@@ -0,0 +1,168 @@
+is used to ensure that you hold the since of the object you are sinking to some
+fixed point
+
+With this refactor, preparation becomes totally unnecessary because we will no
+longer prepare the export, we will only create it under the assumption that it
+will either connect or error.
+
+### Check errors in purification
+
+Just as we perform asynchronous work for sources (e.g. checking that the PG
+publication exists), we should do the same thing for sinks. This gives us a
+chance to error if we believe that the connection will not succeed. Crucially,
+purification should not be side-effecting, so this operation should not leave
+any state behind.
+
+### Remove `SinkConnectionReady`
+
+We will describe connections to Kafka brokers where we plan to write the sink,
+but will no longer "build" the connections in the controller. This means that
+`SinkConnectionReady` will be obviated as `prepare_export` is.
+
+## Alternatives
+
+We could, theoretically, treat `CreateExportToken` as a proxy for the sink in
+reconciliation, which would let us understand that a sink will be created with
+some given ID, even if it is not yet present. However, I am leery of this
+approach because of its complexity (how far down do you propagate
+`CreateExportToken`s? does it differ during boostrapping vs. sink creation?) and
+its lack of parallelism with sources + the idioms of SQL planning.
+
+## Open questions
+
+### Unexpectedly deleted topics
+
+tl;dr––How do we feel about automatically recreating a sink's topic if the topic
+gets deleted?
+
+In the design doc in this PR, I planned on creating the Kafka broker collateral
+(the sink and progress topics) when rendering the sink's dataflow. This design
+is compelling for a few reasons:
+
+-   It's the first time after purification we're permitted to execute
+    asynchronous code (i.e. it cannot be executed before)
+-   It's the first time we're permitted to execute side-effecting code
+
+I have [a draft PR](https://github.com/MaterializeInc/materialize/pull/21852)
+that uses this approach and it's sufficient to handle all expected use cases
+_except_ for the Kafka topic getting unexpectedly deleted while the sink is
+running.
+
+In the case where the sink's topic gets deleted, the sink dataflow tries to
+produce records to its topic, encounters an error, and then halts. When the
+sink's dataflow eventually re-renders, we will re-create the dropped
+collateral––unaware that the user deleted it.
+
+_note_: In the current version of the code, the halting error triggers a
+SuspendAndRestart, though even if we managed to put the sink into a Failing
+state, a reboot of environmentd and the cluster would be sufficient to re-render
+the dataflow.
+
+A short question we need to answer is, "Is it OK to re-create the sink's deleted
+topic?" My gut says that having a zombie producer that doesn't honor a topic's
+deletion is either bad or unexpected. I asked Ben about this, and he had the
+opposite reaction––this is probably fine.
```

Comment:
I'm also in favor of not recreating the Kafka topic if the progress topic already has a progress entry for the sink. Agree we needn't spend too much time handling "out of spec" situations gracefully, but recreating the Kafka topic after it's deleted is going to be fairly surprising for users, and so during a quick check of the progress topic to avoid the weirdness seems well worth it if it's simple.

Reply:
Ah, I had not realized that all sinks that use the same connection use the same progress topic, so there is no way to determine if a missing topic was one that we had previously produced to or not.

I am unsure of how progress topics work by and large, but can definitively say that they contain no data about which exports the encoded timestamps pertain to.

As part of closing out this work, I'll file a ticket that this behavior is undesirable and we should look into ways of improving it.

Reply:
> so there is no way to determine if a missing topic was one that we had previously produced to or not.

True in general, but we should be able to tell if a particular _sink_ has produced data in the past! (Happy to grab some time if you want to brainstorm on this live.)

Reply:
Yep––comment here was wrong. I was unintentionally only decoding the value and not the key and the key contains the data I actually need.